### PR TITLE
Telcodocs 1771 Extend workload partitioning to support cpu limits

### PR DESCRIPTION
**Pull request by kquinn1204**
[Telcodocs 1771]:  Extend workload partitioning to support cpu limits

Version(s): 4.16, main

Issue:https://issues.redhat.com/browse/TELCODOCS-1771 

Link to docs preview: https://75078--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/enabling-workload-partitioning.html

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/enabling-workload-partitioning.adoc [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Module included in the following assemblies:
+// 
+// 
+
+[id="enabling-workload-partitioning_{context}"]
+=  Enabling workload workload partitioning 
+
+With workload partitioning, you annotate the set of cluster management pods and a set of typical add-on Operators for inclusion in the cluster management workload partition. These pods operate normally within the minimum size CPU configuration. Additional Operators or workloads outside of the set of minimum cluster management pods require additional CPUs to be added to the workload partition.
```
Comment (djuran) [inline modules/enabling-workload-partitioning.adoc:8]: Suggest to re-phrase the part about `With workload partitioning, you annotate the set of cluster management pods`
Current phrasing makes it sound like it's the end user who is supposed to add these annotations, while in reality the annotations are there as part of the Red Hat product.

Also the part about "Additional Operators and workloads" is a bit strange.
Which is the initial set of operators and workloads?

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 2]
```diff
@@ -6,57 +6,34 @@ include::_attributes/common-attributes.adoc[]
 
 toc::[]
 
-In resource-constrained environments, you can use workload partitioning to isolate {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs.
+Workload partitioning segregates particular workloads into distinct CPU sets. The primary objective is ensure that the platform pods will not exceed a specified CPU usage limit. You can use workload partitioning to isolate {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in their deployment remain untouched and available exclusively for their own workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).
```
Comment (djuran) [inline scalability_and_performance/enabling-workload-partitioning.adoc:9]: It's not the workloads that get segregated but rather the compute node cpu resources.

And the objective is not to ensure that platform pods don't exceed usage limits but rather that system tasks are restricted to specific CPUs only. And also what you mention below, that the cpu capacity of the host is correctly accounted for

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 3]
```diff
@@ -6,57 +6,34 @@ include::_attributes/common-attributes.adoc[]
...
-+
-[NOTE]
-====
-Workload partitioning can only be enabled during cluster installation. You cannot disable workload partitioning postinstallation.
-====
-
-. In the performance profile, specify the `isolated` and `reserved` CPUs.
-+
-.Recommended performance profile configuration
+If, for example, a request is set to 20 and the limit to 40, the scheduler prioritizes the lower value to allocate resources according to actual usage. Requesting 20 represents the typical need, while 40 is the upper limit. Prioritizing requests over limits optimizes resource allocation, aligning with usage needs and avoiding excess provisioning.
```
Comment (djuran) [inline scalability_and_performance/enabling-workload-partitioning.adoc:17]: Is this only relevant to to platform Pods?
If so, why is this relevant to the end user, they are not supposed to create those

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 4]
```diff
@@ -6,57 +6,34 @@ include::_attributes/common-attributes.adoc[]
...
+In the context of enabling workload partitioning and managing CPU resources effectively, nodes within the cluster must identify themselves as supporting a new resource type called `management.workload.openshift.io/cores` and accurately report their CPU capacity. Workload partitioning can be enabled during cluster installation only by adding the additional field: `cpuPartitioningMode` to the `install-config.yaml` file.
 
-Workload partitioning isolates user workloads from platform workloads using standard Kubernetes scheduling capabilities.
+When workload partitioning is enabled, the `management.workload.openshift.io/cores` resource allows the scheduler to correctly assign pods based on the cpushares capacity of the host, not just the default cpuset. This ensures more precise allocation of resources for workload partitioning scenarios.
 
-The following changes are required for workload partitioning:
+Workload partitioning respects CPU requests and limits defined on pods. From 4.16 accurate CPU usage limitations are set for platform pods utilizing CPU partitioning. Resource requests, rather than limits, are the primary value in setting accurate CPU usage limitations.
```
Comment (eggfoobar) [inline scalability_and_performance/enabling-workload-partitioning.adoc:15]: Maybe rework this paragraph a bit,

Something like this, maybe?


> Workload partitioning respects CPU requests and limits defined in the pod spec. As of 4.16 accurate CPU usage limits are set for platform pods utilizing CPU partitioning. Since workload partitioning uses the custom resource type of `management.workload.openshift.io/cores`,  the values for `requests` and `limits` will equal the same due to a requirement by Kubernetes for extended resources. However, the annotations modified by workload partitioning will correctly reflect the desired limits.

File: scalability_and_performance/enabling-workload-partitioning.adoc (hunk 3 above)
Comment (eggfoobar) [reply scalability_and_performance/enabling-workload-partitioning.adoc:17]: This might be too much detail, we're essentially describing how the scheduler works here. 

The only thing I would want to make mention of is something I wrote above regarding the custom resource values equaling the same for requests and limits. That's a requirement of Kubernetes

https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#consuming-extended-resources
> Note: Extended resources cannot be overcommitted, so request and limit must be equal if both are present in a container spec.

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 5]
```diff
@@ -6,57 +6,37 @@ include::_attributes/common-attributes.adoc[]
 
 toc::[]
 
-In resource-constrained environments, you can use workload partitioning to isolate {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs.
+Workload partitioning segregates compute node CPU resources into distinct CPU sets. The primary objective is to keep platform pods on the specified cores to avoid interrupting the CPUs the customer workloads are running on. Workload partitioning isolates {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in their deployment remain untouched and available exclusively for their own workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).
```
Comment (eggfoobar) [inline scalability_and_performance/enabling-workload-partitioning.adoc:9]: Small nit for the `This ensures that the remaining CPUs in their deployment remain untouched and available exclusively for their own workloads.` sentence. Maybe something like the one below is a little clearer wdyt?

>This ensures that the remaining CPUs in the cluster deployment are untouched and available exclusively for non platform workloads.

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 6]
```diff
@@ -6,57 +6,37 @@ include::_attributes/common-attributes.adoc[]
...
 
-The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).
-With workload partitioning, you annotate the set of cluster management pods and a set of typical add-on Operators for inclusion in the cluster management workload partition.
-These pods operate normally within the minimum size CPU configuration.
-Additional Operators or workloads outside of the set of minimum cluster management pods require additional CPUs to be added to the workload partition.
+In the context of enabling workload partitioning and managing CPU resources effectively, nodes within the cluster must identify themselves as supporting a new resource type called `management.workload.openshift.io/cores` and accurately report their CPU capacity. Workload partitioning can be enabled during cluster installation only by adding the additional field: `cpuPartitioningMode` to the `install-config.yaml` file.
```
Comment (eggfoobar) [inline scalability_and_performance/enabling-workload-partitioning.adoc:11]: We should change this a bit so customers know what is being done automatically and what they might need to be aware of regarding the nodes.

Something like this maybe?

>In the context of enabling workload partitioning and managing CPU resources effectively, nodes that are not configured correctly will not be allowed to join the cluster via a node admission webhook. When the workload partitioning feature is enabled, the Machine Config Pools for `master` and `worker` will be supplied with configurations for nodes to use. Adding new nodes to these pools will make sure they are correctly configured before joining the cluster. 
>
>Currently, it is required that nodes are homogenous per Machine Config Pool in order to ensure correct CPU Affinity is set across all nodes with in that pool. Once admitted nodes within the cluster will identify themselves as supporting a new resource type called `management.workload.openshift.io/cores` and accurately report their CPU capacity. Workload partitioning can be enabled during cluster installation only by adding the additional field: `cpuPartitioningMode` to the `install-config.yaml` file.

**Comment by kquinn1204** [comment]
@djuran @eggfoobar are you happy with the changes I have made if so please add a LGTM

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 7]
```diff
@@ -6,57 +6,39 @@ include::_attributes/common-attributes.adoc[]
 
 toc::[]
 
-In resource-constrained environments, you can use workload partitioning to isolate {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs.
+Workload partitioning segregates compute node CPU resources into distinct CPU sets. The primary objective is to keep platform pods on the specified cores to avoid interrupting the CPUs the customer workloads are running on. Workload partitioning isolates {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in the cluster deployment are untouched and available exclusively for non platform workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).
```
Comment (adellape) [inline scalability_and_performance/enabling-workload-partitioning.adoc:9]: * See ISG on avoiding "segregate". 
* This first paragraph felt pretty big, so suggest breaking it up somehow.
* Re: "non platform" - ISG seems to suggest no space or hyphen for the "non" prefix. Personally "nonplatform" feels weird to me. Maybe "non-platform"? Either way, something other than a space.
```suggestion
Workload partitioning separates compute node CPU resources into distinct CPU sets. The primary objective is to keep platform pods on the specified cores to avoid interrupting the CPUs the customer workloads are running on. 

Workload partitioning isolates {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in the cluster deployment are untouched and available exclusively for non-platform workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).
```

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 8]
```diff
@@ -6,57 +6,39 @@ include::_attributes/common-attributes.adoc[]
...
 
-The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).
-With workload partitioning, you annotate the set of cluster management pods and a set of typical add-on Operators for inclusion in the cluster management workload partition.
-These pods operate normally within the minimum size CPU configuration.
-Additional Operators or workloads outside of the set of minimum cluster management pods require additional CPUs to be added to the workload partition.
+In the context of enabling workload partitioning and managing CPU resources effectively, nodes that are not configured correctly will not be permitted to join the cluster through a node admission webhook. When the workload partitioning feature is enabled, the Machine Config Pools for master and worker will be supplied with configurations for nodes to use. Adding new nodes to these pools will make sure they are correctly configured before joining the cluster.
```
Comment (adellape) [inline scalability_and_performance/enabling-workload-partitioning.adoc:11]: * Re: "master", can it be "control plane" here? Per [SSG](https://redhat-documentation.github.io/supplementary-style-guide/#_master_and_slave) and [repo term glossary](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#control-plane).
* I think "machine config pool" is mostly lowercased throughout the repo.

```suggestion
In the context of enabling workload partitioning and managing CPU resources effectively, nodes that are not configured correctly will not be permitted to join the cluster through a node admission webhook. When the workload partitioning feature is enabled, the machine config pools for control plane and worker will be supplied with configurations for nodes to use. Adding new nodes to these pools will make sure they are correctly configured before joining the cluster.
```

**Review by adellape — COMMENTED** [review]
Reminder for squashing at the end, too.

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 9]
```diff
@@ -6,57 +6,39 @@ include::_attributes/common-attributes.adoc[]
...
+
+
+.Sample performance profile configuration
 [source,yaml]
 ----
 include::snippets/ztp_PerformanceProfile.yaml[]
 ----
-+
+
 include::snippets/performance-profile-workload-partitioning.adoc[]
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply scalability_and_performance/enabling-workload-partitioning.adoc:40]: I added like this to really promote using the PPC tool as a way to create a Performance profile, I am giving this greater prominence based on dialog with QE. Keeping the snippets as reference really hand crafting a perf profile is tricky.

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 10]
```diff
@@ -6,57 +6,41 @@ include::_attributes/common-attributes.adoc[]
...
-+
-.Recommended performance profile configuration
+include::modules/enabling-workload-partitioning.adoc[leveloffset=+1]
+
+include::modules/create-perf-profile-workload-partitioning.adoc[leveloffset=+1]
+
+[role="_additional-resources"]
+.Additional resources
+
+* For more information about creating a performance profile, see xref:../scalability_and_performance/cnf-create-performance-profiles.adoc#cnf-create-performance-profiles[Creating a performance profile].
```
Comment (aireilly) [inline scalability_and_performance/enabling-workload-partitioning.adoc:33]:
```suggestion
* For more information about creating a performance profile, see xref:../edge_computing/ztp-advanced-policy-config.adoc#cnf-create-performance-profiles[Creating a performance profile].
.
```

File: scalability_and_performance/enabling-workload-partitioning.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (adellape) [reply scalability_and_performance/enabling-workload-partitioning.adoc]: Latest arrangement makes sense to me! 👍

**Review by adellape — COMMENTED** [review]
Updates LGTM, noticed a couple final nits I missed earlier.

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 11]
```diff
@@ -6,59 +6,42 @@ include::_attributes/common-attributes.adoc[]
...
-+
-.Recommended performance profile configuration
+include::modules/enabling-workload-partitioning.adoc[leveloffset=+1]
+
+include::modules/create-perf-profile-workload-partitioning.adoc[leveloffset=+1]
+
+[role="_additional-resources"]
+.Additional resources
+
+* xref:../scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc#cnf-about-the-profile-creator-tool_cnf-low-latency-perf-profile[About the Performance Profile Creator].
```
Comment (adellape) [inline scalability_and_performance/enabling-workload-partitioning.adoc:33]: Drop the period here:
```suggestion
* xref:../scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc#cnf-about-the-profile-creator-tool_cnf-low-latency-perf-profile[About the Performance Profile Creator]
```

File: scalability_and_performance/enabling-workload-partitioning.adoc [hunk 12]
```diff
@@ -6,59 +6,42 @@ include::_attributes/common-attributes.adoc[]
...
-Workload partitioning introduces an extended `management.workload.openshift.io/cores` resource type for platform pods.
-kubelet advertises the resources and CPU requests by pods allocated to the pool within the corresponding resource.
-When workload partitioning is enabled, the `management.workload.openshift.io/cores` resource allows the scheduler to correctly assign pods based on the `cpushares` capacity of the host, not just the default `cpuset`.
+include::snippets/performance-profile-workload-partitioning.adoc[]
 
 [role="_additional-resources"]
 .Additional resources
 
-* For the recommended workload partitioning configuration for {sno} clusters, see xref:../edge_computing/ztp-reference-cluster-configuration-for-vdu.adoc#ztp-sno-du-enabling-workload-partitioning_sno-configure-for-vdu[Workload partitioning].
+* For the recommended workload partitioning configuration for {sno} clusters, see xref:../edge_computing/ztp-reference-cluster-configuration-for-vdu.adoc#ztp-sno-du-enabling-workload-partitioning_sno-configure-for-vdu[Recommended single-node OpenShift cluster configuration for vDU application workloads -> Workload partitioning].
```
Comment (adellape) [inline scalability_and_performance/enabling-workload-partitioning.adoc:47]: Same here:
```suggestion
* For the recommended workload partitioning configuration for {sno} clusters, see xref:../edge_computing/ztp-reference-cluster-configuration-for-vdu.adoc#ztp-sno-du-enabling-workload-partitioning_sno-configure-for-vdu[Recommended single-node OpenShift cluster configuration for vDU application workloads -> Workload partitioning]
```

File: modules/create-perf-profile-workload-partitioning.adoc [hunk 13]
```diff
@@ -0,0 +1,9 @@
+:_mod-docs-content-type: CONCEPT
+[id="performance-profile-workload-partitioning_{context}"]
+= Performance profiles and workload partitioning
```
Comment (adellape) [inline modules/create-perf-profile-workload-partitioning.adoc:3]: New modules still missing the list of related assemblies in [metadata](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#module-file-metadata):
```suggestion
// Module included in the following assemblies:
//
// * scalability_and_performance/enabling-workload-partitioning.adoc

:_mod-docs-content-type: CONCEPT
[id="performance-profile-workload-partitioning_{context}"]
= Performance profiles and workload partitioning
```

File: modules/enabling-workload-partitioning.adoc [hunk 14]
```diff
@@ -0,0 +1,38 @@
+:_mod-docs-content-type: PROCEDURE
+[id="enabling-workload-partitioning_{context}"]
+= Enabling workload partitioning
```
Comment (adellape) [inline modules/enabling-workload-partitioning.adoc:3]: Same here:
```suggestion
// Module included in the following assemblies:
//
// * scalability_and_performance/enabling-workload-partitioning.adoc

:_mod-docs-content-type: PROCEDURE
[id="enabling-workload-partitioning_{context}"]
= Enabling workload partitioning 
```

**Review by adellape — COMMENTED** [review]
Last round then I think ready for merge:

**Comment by kquinn1204** [comment]
Hopefully that is it now @adellape sorry for all messing around end of week and all that. Thanks

**Review by adellape — APPROVED** [review]

**Review by adellape — APPROVED** [review]
